yt_dlp/YoutubeDL.py

```diff
@@ -2153,23 +2155,42 @@ def _check_formats(self, formats):
                     except OSError:
                         self.report_warning('Unable to delete temporary file "%s"' % temp_file.name)
             if success:
+                f['__working'] = True
                 yield f
             else:
                 self.to_screen('[info] Unable to download format %s. Skipping...' % f['format_id'])
 
+    def _select_formats(self, formats, selector):
+        return list(selector({
+            'formats': formats,
+            'has_merged_format': any('none' not in (f.get('acodec'), f.get('vcodec')) for f in formats),
+            'incomplete_formats': (all(f.get('vcodec') == 'none' for f in formats)  # No formats with video
+                                   or all(f.get('acodec') == 'none' for f in formats)),  # OR, No formats with audio
+        }))
+
     def _default_format_spec(self, info_dict, download=True):
 
         def can_merge():
             merger = FFmpegMergerPP(self)
             return merger.available and merger.can_merge()
 
+        def evaluate_formats(format_spec):
+            formats = self._get_formats(info_dict)
+            return self._select_formats(formats, self.build_format_selector(format_spec))
+
+        to_stdout = self.params['outtmpl']['default'] == '-'
+
+        if not can_merge() and not to_stdout and evaluate_formats('b/bv+ba') != evaluate_formats('bv*+ba/b'):
+            self.report_warning('ffmpeg not found. The downloaded format is not the highest available quality. '
+                                'Installing ffmpeg is strongly recommended: https://github.com/yt-dlp/yt-dlp#dependencies')
+
         prefer_best = (
             not self.params.get('simulate')
             and download
```

Comment:
Side note: I think having `simulate` result in a different default format spec should be considered a bug. A dry-run should be no different than the real thing, right?

It looks like we can trace this logic back to https://github.com/ytdl-org/youtube-dl/commit/0017d9ad6de831384e74db14a821e4c94020c9ac

Back then, upstream's default format spec was only `best` if ffmpeg was not available. So a simulated run would result in a "requested formats not available" error if ffmpeg was not available and there was no combined video+audio format available. This `simulate` checks seems to be added so that you wouldn't have to manually pass `-f bv+ba` or `-f bv` etc in this scenario -- see the linked issue

(sepro and I started digging into this when I was questioning if `simulate` should result in a warning or not)

Reply:
> (sepro and I started digging into this when I was questioning if `simulate` should result in a warning or not)

And if we don't change the default format spec behavior with `simulate`, then currently the warning is wrongly raised since we aren't preferring best

Reply:
I don't remember anyone having complained, and I am hesitant to change long standing behavior without reason

> And if we don't change the default format spec behavior with `simulate`, then currently the warning is wrongly raised since we aren't preferring best

Edited https://github.com/yt-dlp/yt-dlp/pull/9805#issuecomment-2091273363 to address this

Reply:
Well, consider this me complaining lol
As bash mentioned, I think a simulate/dry run option is only really useful if you actually simulate what will happen. If a user wants to see what formats will be selected when downloading, the result would be wrong without FFmpeg installed.
This behaviour is currently also not documented.

Reply:
I was bitten by this undocumented behavior the other day when I was troubleshooting for someone on discord. I also would like to see it fixed or at the very least documented

Reply:
Still, I don't want to make this change in this PR. Could you open a new issue? We should discuss with dirkf as well.